tests/countries/test_sao_tome_and_principe.py

```diff
@@ -0,0 +1,126 @@
+#           ryanss <ryanssdev@icloud.com> (c) 2014-2017
+#  Website: https://github.com/vacanza/holidays
+#  License: MIT (see LICENSE file)
+
+from unittest import TestCase
+
+from holidays.countries.sao_tome_and_principe import SaoTomeAndPrincipe
+from tests.common import CommonCountryTests
+
+
+class TestSaoTomeAndPrincipe(CommonCountryTests, TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass(SaoTomeAndPrincipe, years=range(2014, 2050))
+        cls.observed_holidays = SaoTomeAndPrincipe(years=range(2014, 2050), observed=True)
+
+    def test_new_years_day(self):
+        for year in range(2014, 2050):
+            self.assertHolidayName("Ano Novo", f"{year}-01-01")
+
+    def test_king_amador_day(self):
+        for year in range(2014, 2050):
+            self.assertHolidayName("Dia do Rei Amador", f"{year}-01-04")
+
+    def test_martyrs_day(self):
+        for year in range(2014, 2050):
+            self.assertHolidayName("Dia dos Mártires", f"{year}-02-03")
+
+    def test_labor_day(self):
+        for year in range(2014, 2050):
+            self.assertHolidayName("Dia do Trabalhador", f"{year}-05-01")
+
+    def test_independence_day(self):
+        for year in range(2014, 2050):
+            self.assertHolidayName("Dia da Independência", f"{year}-07-12")
+
+    def test_armed_forces_day(self):
+        for year in range(2014, 2050):
+            self.assertHolidayName("Dia das Forças Armadas", f"{year}-09-06")
+
+    def test_agrarian_reform_day(self):
+        for year in range(2014, 2050):
+            self.assertHolidayName("Dia da Reforma Agrária", f"{year}-09-30")
+
+    def test_sao_tome_day(self):
+        for year in range(2019, 2050):  # Only from 2019 onward
+            self.assertHolidayName("Dia de São Tomé", f"{year}-12-21")
+
+    def test_christmas_day(self):
+        for year in range(2014, 2050):
+            self.assertHolidayName("Natal", f"{year}-12-25")
+
+    def test_observed_holidays(self):
+        # Test New Year's Day observed (Sunday -> Monday)
+        self.assertIn("2023-01-02", self.observed_holidays)
+        self.assertEqual(self.observed_holidays["2023-01-02"], "Ano Novo (observado)")
+
+        # Test Independence Day observed (Sunday -> Monday)
+        self.assertIn("2020-07-13", self.observed_holidays)
+        self.assertEqual(self.observed_holidays["2020-07-13"], "Dia da Independência (observado)")
+
+    def test_principe_subdivision(self):
+        principe_holidays = SaoTomeAndPrincipe(subdiv="P", years=range(2020, 2026))
+
+        principe_fixed = [
+            ("01-17", "Descobrimento da Ilha do Príncipe"),
+            ("04-29", "Dia da Autonomia do Príncipe"),
+            ("08-15", "Dia de São Lourenço"),
+        ]
+
+        for month_day, name in principe_fixed:
+            for year in range(2020, 2026):
+                self.assertHolidayName(name, principe_holidays, f"{year}-{month_day}")
```

Comment:
You might want to consider splitting this test case into separate ones, especially if each of Príncipe’s holidays has its own observed variation later:
```suggestion
    def test_discovery_of_principe_island(self):
        name = "Descobrimento da Ilha do Príncipe"
        self.assertHolidayName(
            name, subdiv_p_holidays, (f"{year}-01-17" for year in range(2014, 2050))
        )
        self.assertNoHolidayName(name)

    def test_autonomy_day(self):
        name = "Dia da Autonomia do Príncipe"
        self.assertHolidayName(
            name, subdiv_p_holidays, (f"{year}-04-29" for year in range(2014, 2050))
        )
        self.assertNoHolidayName(name)

    def test_sao_lorenco_day(self):
        name = "Dia de São Lourenço"
        self.assertHolidayName(
            name, subdiv_p_holidays, (f"{year}-08-15" for year in range(2014, 2050))
        )
        self.assertNoHolidayName(name)
```